pkg/codegen/go/gen_program_ternaries.go

```diff
@@ -0,0 +1,64 @@
+package gen
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/pulumi/pulumi/pkg/v2/codegen/hcl2/model"
+	"github.com/pulumi/pulumi/pkg/v2/codegen/hcl2/syntax"
+)
+
+type ternaryTemp struct {
+	Name  string
+	Value *model.ConditionalExpression
+}
+
+func (tt *ternaryTemp) Type() model.Type {
+	return tt.Value.Type()
+}
+
+func (tt *ternaryTemp) Traverse(traverser hcl.Traverser) (model.Traversable, hcl.Diagnostics) {
+	return tt.Type().Traverse(traverser)
+}
+
+func (tt *ternaryTemp) SyntaxNode() hclsyntax.Node {
+	return syntax.None
+}
+
+type tempAllocator struct {
```

Comment:
Nit: technically, what we're doing is spilling, so I'd name this `tempSpiller`
```suggestion
type tempSpiller struct {
```